src/tools/sc-hsm-tool.c

```diff
@@ -1667,6 +1693,322 @@ static int unwrap_key(sc_card_t *card, int keyid, const char *inf, const char *p
+	size_t datalen, outlen;
+
+	if ((keyid < 1) || (keyid > 255)) {
+		fprintf(stderr, "Invalid key reference (must be 0 < keyid <= 255)\n");
+		return -1;
+	}
+
+	fid[0] = EE_CERTIFICATE_PREFIX;
+	fid[1] = (unsigned char)keyid;
+	ef_cert_len = 0;
+
+	/* Try to select a related EF containing the certificate for the key */
+	sc_path_set(&path, SC_PATH_TYPE_FILE_ID, fid, sizeof(fid), 0, 0);
+	r = sc_select_file(card, &path, NULL);
+	if (r != SC_SUCCESS) {
+		fprintf(stderr, "Wrong key reference (-i %d)? Failed to select file: %s\n", keyid, sc_strerror(r));
+		return -1;
+	}
+
+	ef_cert_len = sc_read_binary(card, 0, ef_cert, sizeof(ef_cert), 0);
+	if (ef_cert_len < 0) {
+		fprintf(stderr, "Error reading certificate %s. Skipping\n", sc_strerror(ef_cert_len));
+		ef_cert_len = 0;
+	} else {
+		ef_cert_len = determineLength(ef_cert, ef_cert_len);
+	}
+
+	/* C_DevAut */
+	fid[0] = 0x2F;
+	fid[1] = 0x02;
+	dev_aut_cert_len = 0;
+
+	/* Read concatenation of both certificates */
+	sc_path_set(&path, SC_PATH_TYPE_FILE_ID, fid, sizeof(fid), 0, 0);
+	r = sc_select_file(card, &path, NULL);
+	if (r != SC_SUCCESS) {
+		fprintf(stderr, "Failed to select certificates: %s\n", sc_strerror(r));
+		return -1;
+	}
+
+	total_certs_len = sc_read_binary(card, 0, dev_aut_cert, sizeof(dev_aut_cert), 0);
+	if (total_certs_len < 0) {
+		fprintf(stderr, "Error reading certificate: %s\n", sc_strerror(total_certs_len));
+		return -1;
+	} else {
+		dev_aut_cert_len = determineLength(dev_aut_cert, total_certs_len);
+		dica_len = total_certs_len - dev_aut_cert_len;
+		memcpy(dica, dev_aut_cert + dev_aut_cert_len, dica_len);
+	}
+	if (dica_len == 0) {
+		fprintf(stderr, "Could not determine device issuer certificate\n");
+		return -1;
+	}
+
+	if ((outfp = fopen(outf, "r"))) {
+		fprintf(stderr, "Output file '%s' already exists\n", outf);
+		fclose(outfp);
+		return -1;
+	}
+	fprintf(stderr, "Warning: Device certificate chain not verified!\n");
+
+	datalen = ef_cert_len + dev_aut_cert_len + dica_len;
+	outlen = 8 + datalen;
+	if (!(data = malloc(datalen))) {
+		fprintf(stderr, "Malloc failed\n");
+		r = -1;
+		goto err;
+	}
+	if (!(out = malloc(outlen))) {
+		fprintf(stderr, "Malloc failed\n");
+		r = -1;
+		goto err;
+	}
+	memcpy(data, ef_cert, ef_cert_len);
+	memcpy(data + ef_cert_len, dev_aut_cert, dev_aut_cert_len);
+	memcpy(data + ef_cert_len + dev_aut_cert_len, dica, dica_len);
+
+	if ((r = sc_asn1_put_tag(tag, data, datalen, out, outlen, &ptr)) < 0) {
+		fprintf(stderr, "Error formatting ASN1 sequence: %s\n", sc_strerror(r));
+		r = -1;
+		goto err;
+	}
+	outlen = ptr - out;
+
+	if (!(outfp = fopen(outf, "wb"))) {
+		perror(outf);
+		r = -1;
+		goto err;
+	}
+
+	if (fwrite(out, 1, outlen, outfp) != (size_t)outlen) {
+		perror(outf);
+		r = -1;
+		goto err;
+	}
+
+err:
+	if (outfp)
+		fclose(outfp);
+	if (out)
+		free(out);
+	if (data)
+		free(data);
+
+	return r;
+}
+
+static void get_CHR(char *chrstr, int is_cvc, sc_context_t *ctx, const u8 *buf, size_t buflen)
```

Comment:
please use sc_pkcs15emu_sc_hsm_decode_cvc for parsing

Reply:
Regarding EF_C_DevAut I can see that it is set in the static function `sc_hsm_perform_chip_authentication()`. `sc_hsm_perform_chip_authentication()` is only called from `sc_hsm_pin_cmd()`. That is, filling EF_C_DevAut currently requires PIN entry, which is not necessary for my implementation.

What would be a good way to use EF_C_DevAut, ideally without PIN entry?

Reply:
Regarding `sc_pkcs15emu_sc_hsm_decode_cvc()`: I spent many hours trying to get it to work.
Like I wrote in https://github.com/OpenSC/OpenSC/pull/1711#issuecomment-546378453, I wasn't able to parse the device certificate and the device issuer certificate with it. For the public key, `sc_cvc_t.car` is filled correctly, but `sc_cvc_t.outer_car` contains garbage. Unfortunately, I couldn't figure out what the problem is.

Reply:
> What would be a good way to use EF_C_DevAut, ideally without PIN entry?

The basic idea is in sc-hsm-tool to
1. force a sc-hsm card driver, as done [here](https://github.com/OpenSC/OpenSC/pull/1896/files#diff-1252f4f12c940c3ec84eebedcd816a8fR600-R604) and
2. then cast card->drv_data to `struct sc_hsm_private_data *` and use its `EF_C_DevAut` if it's available.

Additionally, you may want to transfer ownership of a newly read EF_C_DevAut to the driver for re-use. You don't need to perform CA to do this.